### PR TITLE
[WIP] Integration tests for submission worker.

### PR DESCRIPTION
**Pull request by isht3**
This is how I had an idea for testing the submission worker. We could probably use https://code.google.com/archive/p/mock/ to mock RabbitMQ but that is up for discussion. Also, the other way would be to use `subprocess` module to actually run the `submission_worker`. This is just a hint of the implementation, if we decide to stick to this, I'll start with a proper one.

### CODE REVIEW

File: settings/common.py [hunk 1]
```diff
@@ -283,3 +283,5 @@
 REST_AUTH_SERIALIZERS = {
     'USER_DETAILS_SERIALIZER': 'accounts.serializers.ProfileSerializer',
 }
+
+TEST = 'test' in sys.argv
```
Comment (deshraj) [inline settings/common.py]: Please replace the variable name with `TESTING`

**Comment by isht3** [comment]
@deshraj Okay, thanks. I'll set up a dummy server using a mock server.

**Comment by isht3** [comment]
@deshraj I was thinking about using https://github.com/eandersson/amqpstorm What do you think?

**Comment by deshraj** [comment]
Is there a specific reason why you want to go with this instead of mock (the previous library that you suggested)?

**Comment by isht3** [comment]
I found https://www.amqpstorm.io/ to have more straightforward documentation, but I haven't tested it yet. Is it okay if I test it and then fix on which one to use?

**Comment by isht3** [comment]
Also, what do you think about splitting the directory of the submission worker for development and production environment?

**Comment by deshraj** [comment]
@isht3: I would not recommend trying it since it will definitely take a lot of time of yours and we might end up not using it (your time is valuable).

**Comment by isht3** [comment]
Okay, thanks. I will then get started with the integration tests! :smile:

**Comment by isht3** [comment]
@deshraj, right now, since the submission worker is designed to run like `python scripts/workers/submission_worker.py` I think we will need to make the submission worker a bit more modular using classes if integration tests are to be written in the given way. Is it okay if I do that? Otherwise, we'd have to use Subprocess, which will become obsolete when we switch to SQS.

**Comment by isht3** [comment]
@deshraj I've set up a few helper functions in submission worker to help with the testing, this way I think we could avoid using any unnecessary outside libraries for testing the submission worker components. Do let me know what you think.

**Comment by isht3** [comment]
@deshraj @RishabhJain2018 
`ImportError: No module named challenge_data.challenge_2` 
I was getting this error from this line
 `challenge_module = importlib.import_module(CHALLENGE_IMPORT_STRING.format(challenge_id=challenge.id))` and I don't quite understand it properly. Can you tell me what that line in currently doing? Any idea what's wrong or how it could be solved?

**Comment by isht3** [comment]
@deshraj Also, is there any particular reason we use postgres database for testing? Can we switch to sqlite3 for testing the submission worker? Because it is leading to a very popular error of not detecting the test database. The preferred solution is switching to sqlite3 while running test.

**Comment by deshraj** [comment]
@isht3: We cannot switch to the sqlite3 database for testing since we are using Django's JSONField which is only supported by PostgreSQL database.

**Comment by deshraj** [comment]
> @deshraj @RishabhJain2018
ImportError: No module named challenge_data.challenge_2
I was getting this error from this line
challenge_module = importlib.import_module(CHALLENGE_IMPORT_STRING.format(challenge_id=challenge.id)) and I don't quite understand it properly. Can you tell me what that line in currently doing? Any idea what's wrong or how it could be solved?

Can you paste the terminal log when you started the submission worker?

**Comment by isht3** [comment]
@deshraj  I'm not manually starting up the submission worker and using API to call the needed functions.

```
ERROR: test_challenge_submission_for_successful_submission (tests.unit.submission_worker.test_submissions.BaseAPITestClass)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/rustbucket/evalai/tests/unit/submission_worker/test_submissions.py", line 123, in test_challenge_submission_for_successful_submission
    'status': 'submitting', 'input_file': self.input_file}, format="multipart")
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/rest_framework/test.py", line 290, in post
    path, data=data, format=format, content_type=content_type, **extra)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/rest_framework/test.py", line 212, in post
    return self.generic('POST', path, data, content_type, **extra)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/django/test/client.py", line 409, in generic
    return self.request(**r)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/rest_framework/test.py", line 279, in request
    return super(APIClient, self).request(**kwargs)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/rest_framework/test.py", line 231, in request
    request = super(APIRequestFactory, self).request(**kwargs)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/django/test/client.py", line 494, in request
    six.reraise(*exc_info)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/django/core/handlers/exception.py", line 39, in inner
    response = get_response(request)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/rest_framework/views.py", line 477, in dispatch
    response = self.handle_exception(exc)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/rest_framework/views.py", line 437, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/rest_framework/views.py", line 474, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/rustbucket/evalai/venv/local/lib/python2.7/site-packages/rest_framework/decorators.py", line 52, in handler
    return func(*args, **kwargs)
  File "/home/rustbucket/evalai/apps/jobs/views.py", line 116, in challenge_submission
    publish_submission_message(challenge_id, challenge_phase_id, submission.id)
  File "/home/rustbucket/evalai/apps/jobs/sender.py", line 23, in publish_submission_message
    body=json.dumps(message))
  File "/home/rustbucket/evalai/tests/unit/submission_worker/fake_publisher.py", line 28, in basic_publish
    raise e
ImportError: No module named challenge_data.challenge_1

```

**Pull request closed without merging**